### PR TITLE
Remove deprecated reverse option from sorting

### PR DESCRIPTION
**Pull request by MaineC**
This removes the "reverse" option for sorting as discussed in #17047 - instead define sort order explicitly like shown here: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-sort.html#_sort_order

Relates to #17047

Pending: PR to deprecate the option removed above in 2.x

### CODE REVIEW

**Comment by cbuescher** [comment]
I found one mention of the 'reverse' parameter related to sort in `./reference/search/request/sort.asciidoc` which should be removed. Maybe also add a not to the migrate.asciidoc, even though the option will be deprecated in 2.x. Other than that, LGTM

**Comment by s1monw** [comment]
can we add a test that makes sure we fail is somebody uses the option?

**Comment by MaineC** [comment]
Thanks for the speedy comments.

* @cbuescher wrt. sort in docs: Will remove the mention, same for adding to migrate.asciidoc
* @s1monw wrt. to adding said test: Makes sense, will do

**Comment by MaineC** [comment]
Updated. Not sure about the change to GeoDistanceSortBuilder - added an explicit check for the token "reverse", without it parsing fell through to [parsing the reference geo point for some field from a geohash string](https://github.com/elastic/elasticsearch/blob/a3fc4c0370671b6d193a6e0887108179f1fc5432/core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java#L471).

Would like to dig a bit more to confirm this behaviour outside the unit test with a real curl reproduction.

**Comment by MaineC** [comment]
This is what we get when a user includes the reverse sort option after removing its support in GeoDistanceSortBuilder:

    {
       "error": {
          "root_cause": [
             {
                "type": "parse_exception",
                "reason": "illegal latitude value [269.97802734375] for [GeoDistanceSort]"
             }
          ],    
          "type": "search_phase_execution_exception",
          "reason": "all shards failed",
          "phase": "query",
          "grouped": true,
          "failed_shards": [
             {
                "shard": 0,
                "index": "products",
                "node": "MeeEqVKMQF2UvjooWHAGlw",
                "reason": {
                   "type": "parse_exception",
                   "reason": "illegal latitude value [269.97802734375] for [GeoDistanceSort]"
                }
             }
          ],
          "caused_by": {
             "type": "parse_exception",
             "reason": "illegal latitude value [269.97802734375] for [GeoDistanceSort]"
          }
       },
       "status": 400
    }

When explicitly forbidding the query option like in the change above, we get something that's more readable IMHO while at the same time forbidding the user to have geo fields named "reverse":

    {
       "error": {
          "root_cause": [
             {
                "type": "parsing_exception",
                "reason": "Sort option [reverse] no longer supported.",
                "line": 6,
                "col": 20
             }
          ],
          "type": "parsing_exception",
          "reason": "Sort option [reverse] no longer supported.",
          "line": 6,
          "col": 20
       },
       "status": 400
    }

My personal preference would be to go with the latter option, @cbuescher any opinion?

**Comment by cbuescher** [comment]
@MaineC I see the problem, but I'm on the fence here with treating "reverse" as a special case that cannot be used as a field name for the geo point option. That would require special documentation (all fieldnames allowed except...) and somehow doesn't feel right to me. What if instead to mitigate the problem:
* dissallow setting the field name for the geo points twice and throw an error with the fieldname already set and the different one we are trying to set (that would give errors sth. like `fieldname already set to "my_point" but trying to reset to "reverse") 
* dissallow token value other than VALUE_STRING in the case where we are parsing to geo hash. That would catch the cases where we have `"reverse" : true` in the query.
* Add the fieldname to the error message above (I think that gets triggered in build()), so users get a hint that they are using the wrong field when accidentally specifying `"reverse" : "true"|"false"`

**Comment by MaineC** [comment]
@cbuescher good points, I like your proposals. Changed the code accordingly. Only gotcha:

> Add the fieldname to the error message above (I think that gets triggered in build()), so users get a 
> hint that they are using the wrong field when accidentally specifying "reverse" : "true"|"false"

While the string "false" evaluates to an invalid coordinate, the string "true" seems to evaluate to a valid one. See here: http://www.movable-type.co.uk/scripts/geohash.html

File: core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java [hunk 1]
```diff
@@ -433,6 +434,13 @@ public static GeoDistanceSortBuilder fromXContent(QueryParseContext context, Str
                     nestedFilter = context.parseInnerQueryBuilder();
                 } else {
                     // the json in the format of -> field : { lat : 30, lon : 12 }
+                    if (fieldName != null) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/sort/GeoDistanceSortBuilder.java]: Nit: maybe we could make this `fieldName.equals(currentName) == false`, so this won't fail when same name is used multiple times. Might potentially happen when generating queries?

**Comment by cbuescher** [comment]
Okay, I'd say we have to live with the small gotcha that `"true"` as geohash is valid (nice, Kasachstan ;-)), better than making exceptions on allowed fieldnames only for catching (probably largely unused) deprecated options. Left one last minor nit, otherwise LGTM.

**Comment by MaineC** [comment]
@cbuescher NIT fixed. Merging as soon as tests are done.

**Pull request merged**